Gems/Terrain/Code/Source/EditorComponents/EditorTerrainPhysicsColliderComponent.cpp

```diff
@@ -14,11 +14,121 @@ namespace Terrain
+    {
+        AZStd::shared_lock<AZStd::shared_mutex> lock(m_drawMutex);
+
+        AZ::Aabb boxBounds = AZ::Aabb::CreateNull();
+        AZ::Transform transform = AZ::Transform::CreateIdentity();
+        LmbrCentral::ShapeComponentRequestsBus::Event(
+            GetEntityId(), &LmbrCentral::ShapeComponentRequestsBus::Events::GetTransformAndLocalBounds, transform, boxBounds);
+
+        int32_t gridWidth, gridHeight;
+        m_component.GetHeightfieldGridSize(gridWidth, gridHeight);
+
+        const AZ::Vector2 gridResolution = m_component.GetHeightfieldGridSpacing();
+        auto heights = m_component.GetHeights();
+
+        m_debugQuads.clear();
+        m_debugQuads.reserve((gridWidth - 1) * (gridHeight - 1));
+
+        for (int xIndex = 0; xIndex < gridWidth - 1; xIndex++)
+        {
+            for (int yIndex = 0; yIndex < gridHeight - 1; yIndex++)
+            {
+                const int index0 = yIndex * gridWidth + xIndex;
+                const int index1 = yIndex * gridWidth + xIndex + 1;
+                const int index2 = (yIndex + 1) * gridWidth + xIndex + 1;
+                const int index3 = (yIndex + 1) * gridWidth + xIndex;
+
+                const float x0 = boxBounds.GetMin().GetX() + gridResolution.GetX() * xIndex;
+                const float x1 = boxBounds.GetMin().GetX() + gridResolution.GetX() * (xIndex + 1);
+                const float y0 = boxBounds.GetMin().GetY() + gridResolution.GetY() * (yIndex + 1);
+                const float y1 = boxBounds.GetMin().GetY() + gridResolution.GetY() * yIndex;
+
+                m_debugQuads.push_back(
+                    DebugImageQuad{
+                        AZ::Vector3(x0, y0, heights[index0]),
+                        AZ::Vector3(x1, y0, heights[index1]),
+                        AZ::Vector3(x1, y1, heights[index2]),
+                        AZ::Vector3(x0, y1, heights[index3])
+                    }
+                );
+            }
+        }
+    }
+
+    void EditorTerrainPhysicsColliderComponent::OnTerrainDataChanged([[maybe_unused]] const AZ::Aabb& dirtyRegion, [[maybe_unused]] TerrainDataChangedMask dataChangedMask)
+    {
+        GenerateDebugDrawData();
+    }
+
+    void EditorTerrainPhysicsColliderComponent::DisplayEntityViewport(
+        [[maybe_unused]] const AzFramework::ViewportInfo& viewportInfo, [[maybe_unused]] AzFramework::DebugDisplayRequests& debugDisplay)
+    {
+        if (!m_visibleInEditor && !IsSelected())
+        {
+            return;
+        }
+
+        AZStd::shared_lock<AZStd::shared_mutex> lock(m_drawMutex);
+
+        AZ::Transform worldFromLocalWithUniformScale = m_component.GetHeightfieldTransform();
+        worldFromLocalWithUniformScale.SetUniformScale(worldFromLocalWithUniformScale.GetUniformScale());
+
+        debugDisplay.PushMatrix(worldFromLocalWithUniformScale);
+        debugDisplay.SetColor(m_drawColor);
+
+        for (const auto quad : m_debugQuads)
```

Comment:
This is copying every element from `m_debugQuads` while iterating it. Missing reference.

````
for (const auto& quad : m_debugQuads)
````